Support for an e2e-framework kubetest2 tester (#168)
kubernetes-sigs/e2e-framework: third_party/kubetest2/README.md

# Kubetest2 E2E-Framework Tester

This third-party package implements a Kubernetes-SIGs/[kubetest2](https://github.com/kubernetes-sigs/kubetest2) tester capable of executing tests written using the [e2e-framework](https://github.com/kubernetes-sigs/e2e-framework).

## Kubetest2 installation

> First, you must ensure that you have the `kubtest2` binary installed in the environment where you will run your tests :

Reviewer: ```suggestion
First, you must ensure that you have the `kubetest2` binary installed in the environment where you will run your tests :
```

```
go install sigs.k8s.io/kubetest2/...@latest
```

> See further detail [here](https://github.com/kubernetes-sigs/kubetest2#installation).

## E2E-Framework installation

> To build the e2e-framework tester, you can use the following `go intstall` command:

Reviewer: ```suggestion
To build the e2e-framework tester, you can use the following `go install` command:
```

      --disable-graceful-teardown   Ignore panic recovery while running tests. This will prevent test finish steps from getting executed on panic
      --dry-run                     Run Test suite in dry-run mode. This will list the tests to be executed without actually running them
      --fail-fast                   Fail immediately and stop running untested code
      --feature string              Regular expression to select feature(s) to test
  -h, --help
      --kubeconfig string           Path to a cluster kubeconfig file (optional)
      --labels string               Comma-separated key=value to filter features by labels
      --namespace string            A namespace value to use for testing (optional)
      --packages string             Space-separated packages to test
      --parallel                    Run test features in parallel
      --skip-assessments string     Regular expression to skip assessment(s) to run
      --skip-features string        Regular expression to skip feature(s) to run
      --skip-labels string          Regular expression to skip label(s) to run
      --test-flags string           Space-separated flags applied to 'go test' command
```

> To run a test with kubetest2, you must follow this command format as outline above:

Reviewer: ```suggestion
To run a test with kubetest2, you must follow this command format as outlined above:
```

```
kubetest2 [<deployer>] --test=e2e-framework -- [e2e-framework-flags]
```

Where:

* `deployer` is a binary used to stand up an infrastructure (not provided here).
* `--test=e2e-framework` specifies the tester to used, in this case the e2e-framework tester.
* `e2e-framework-flags` are the list of CLI flags that are passed to the e2e-framework binary.

### Running a simple tests
Let us use `kubetest2` to run a simple test with no deployer and no arguments passed to the test:

```

> kubetest2 noop --test=e2e-framework -- --packages ./simple

Reviewer: ```suggestion
kubetest2 noop --test=e2e-framework -- --packages ./examples/simple
```

```

The previous command will launch kubetest2 without starting a cluster (noop deployer). Note the `--` which indicates the start of arguments which will be passed to the tester (e2e-framework).

If the command is successful, you will see the `kubetest2` logs similar to what is shown:
```
I1211 08:56:43.836568   14380 app.go:61] The files in RunDir shall not be part of Artifacts
I1211 08:56:43.837144   14380 app.go:62] pass rundir-in-artifacts flag True for RunDir to be part of Artifacts
I1211 08:56:43.837188   14380 app.go:64] RunDir for this run: "/home/ubuntu/go/e2e-framework/examples/_rundir/cf42e94d-1c2e-4c6f-9912-2ff948608008"
I1211 08:56:43.860870   14380 app.go:128] ID for this run: "cf42e94d-1c2e-4c6f-9912-2ff948608008"
Running:  go test  ./simple -args
ok  	sigs.k8s.io/e2e-framework/examples/simple	0.363s
```

> You can easily change this command to ask `kubetest` to stand and teardown a Kubernets cluster using kind for instance (kubetest2 support other deployers), as shown:

Reviewer: ```suggestion
You can easily change this command to ask `kubetest2` to stand and teardown a Kubernetes cluster using kind for instance (`kubetest2` also supports other deployers), as shown:
```

```

> $> kubetest2 kind --up --down --test=e2e-framework -- packages .

Reviewer: ```suggestion
$> kubetest2 kind --up --down --test=e2e-framework -- --packages .
```

I1211 09:20:46.640997   19579 up.go:62] Up(): creating kind cluster...
Creating cluster "kind" ...
 • Ensuring node image (kindest/node:v1.25.3) 🖼  ...
 ✓ Ensuring node image (kindest/node:v1.25.3) 🖼
 • Preparing nodes 📦   ...
...
Running:  go test   -args
PASS
ok  	sigs.k8s.io/e2e-framework/kubetest2test	0.307s
I1211 09:23:14.406049   19579 down.go:32] Down(): deleting kind cluster...
Deleting cluster "" ...
```

> It should be noted that e2e-framework offers its own programmatic hooks for more flexible controls of pre- and post-test tasks, such as creating and tearing down named clusters (see the examples folder).

### Passing arguments to e2e-framework tests

> The kubetest2 e2e-framework tester supports the ability to pass CLI arguments to your tests to control settings such ass test execution filters and test behaviors. Recall, in order for your e2e-framework tests to receive CLI arguments, you must programmatically create the test environment to receive its configuration from the command-line:

Reviewer: ```suggestion
The kubetest2 e2e-framework tester supports the ability to pass CLI arguments to your tests to control settings such as test execution filters and test behaviors. Recall, in order for your e2e-framework tests to receive CLI arguments, you must programmatically create the test environment to receive its configuration from the command-line:
```

```
I1211 08:56:43.836568   14380 app.go:61] The files in RunDir shall not be part of Artifacts
I1211 08:56:43.837144   14380 app.go:62] pass rundir-in-artifacts flag True for RunDir to be part of Artifacts
I1211 08:56:43.837188   14380 app.go:64] RunDir for this run: "/home/ubuntu/go/e2e-framework/examples/_rundir/cf42e94d-1c2e-4c6f-9912-2ff948608008"
I1211 08:56:43.860870   14380 app.go:128] ID for this run: "cf42e94d-1c2e-4c6f-9912-2ff948608008"
Running:  go test  ./simple -args
ok  	sigs.k8s.io/e2e-framework/examples/simple	0.363s
```

You can easily change this command to stand up, and teardown, a Kubernetes cluster using KinD for instance (`kubetest2` also supports other deployers):

```
$> kubetest2 kind --up --down --test=e2e-framework -- --packages .
I1211 09:20:46.640997   19579 up.go:62] Up(): creating kind cluster...
Creating cluster "kind" ...
 • Ensuring node image (kindest/node:v1.25.3) 🖼  ...
 ✓ Ensuring node image (kindest/node:v1.25.3) 🖼
 • Preparing nodes 📦   ...
...
Running:  go test   -args
PASS
ok  	sigs.k8s.io/e2e-framework/kubetest2test	0.307s
I1211 09:23:14.406049   19579 down.go:32] Down(): deleting kind cluster...
Deleting cluster "" ...
```

> > It should be noted that e2e-framework offers its own programmatic hooks for more flexible controls of pre- and post-test tasks, such as creating and tearing down named clusters (see the examples folder).

Reviewer 2: ```suggestion
> It should be noted that e2e-framework offers its own programmatic hooks for more flexible controls of pre and post test tasks, such as creating and tearing down named clusters (see the examples folder).
```
Minor nitpick here too.

### Passing arguments to e2e-framework tests

> The kubetest2 e2e-framework tester supports the ability to pass CLI arguments to your tests to control settings such as test execution filters and test behaviors. Recall, in order for your e2e-framework tests to receive CLI arguments, you must programmatically create the test environment to receive its configuration from the command-line:

Reviewer 2: ```suggestion
The kubetest2 e2e-framework tester supports the ability to pass CLI arguments for your tests to control settings such as test execution filters and test behaviors. Recall, in order for your e2e-framework tests to receive CLI arguments, you must programmatically create the test environment to receive its configuration from the command-line:
```

```go

var tenv env.Environment

func TestMain(m *testing.M) {

> 	cfg, err := envconf.NewFromFlags() // important: read config from CLI flags
> 	if err != nil {
> 		log.Fatalf("failed to build config from flags: %s", err)
> 	}
> 	tenv = env.NewWithConfig(cfg)

Reviewer 2: This can be replaced with the new function you have included `NewFromFlags` to make this much easier I think,

```suggestion
        var err error
        tenv, err = env.NewFromFlags()
	if err != nil {
		log.Fatalf("failed to build config from flags: %s", err)
	}
```

	os.Exit(tenv.Run(m))
}

func TestClusterObjects(t *testing.T) {
	f := features.New("cluster-test")
	f.Assess("pod-count", func(ctx context.Context, t *testing.T, cfg *envconf.Config) context.Context {
		var pods corev1.PodList
		client, err := cfg.NewClient()
		if err != nil {
			t.Fatal(err)
		}
		err = client.Resources("kube-system").List(context.TODO(), &pods)
		if err != nil {
			t.Fatal(err)
		}
		if len(pods.Items) == 0 {
			t.Fatal("no pods in namespace kube-system")
		}
		return ctx
	})
	f.Assess("dep-count", func(ctx context.Context, t *testing.T, cfg *envconf.Config) context.Context {
		var deps appsv1.DeploymentList
		client, err := cfg.NewClient()
		if err != nil {
			t.Fatal(err)
		}
		err = client.Resources().List(context.TODO(), &deps)
		if err != nil {
			t.Fatal(err)
		}
		if len(deps.Items) == 0 {

> 			t.Fatal("apps found")

Reviewer 2: Not sure if this error is right. Why does this say `apps found` where the check is `== 0` ?
